Assets/PatchKit Patcher/Scripts/AppUpdater/Commands/CheckVersionIntegrityCommand.cs

```diff
@@ -95,19 +97,80 @@ public override void Execute(CancellationToken cancellationToken)
             }
         }
 
+        private static volatile FileIntegrity[] files;
+
         private void ExecuteInternal(CancellationToken cancellationToken)
         {
             DebugLogger.Log("Checking version integrity.");
 
             _status.IsActive.Value = true;
 
-            var files = new FileIntegrity[_versionSummary.Files.Length];
-
-            for (int i = 0; i < _versionSummary.Files.Length; i++)
+            files = new FileIntegrity[_versionSummary.Files.Length];
+            int i = 0;
+            foreach (var file in _versionSummary.Files)
             {
-                files[i] = CheckFile(_versionSummary.Files[i]);
+                while (true)
+                {
+                    int tmp;
+                    lock (this)
+                    {
+                        tmp = _space;
+                    }
+
+                    if (tmp < 16)
+                    {
+                        break;
+                    }
+
+                    Thread.Sleep(1);
+                }
+
+                lock (this)
+                {
+                    _space++;
+                }
+
+                var fileInThread = file;
+                var number = i;
+                ThreadPool.QueueUserWorkItem(state =>
+                {
+                    try
+                    {
+                        var fileIntegrity = CheckFile(fileInThread);
+                        lock (this)
+                        {
+                            files[number] = fileIntegrity;
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        DebugLogger.LogError(e.ToString());
+                        throw;
```

Comment:
You shouldn't allow a thread to end up with an exception. A different solution is required here (like storing `lastThreadException` that is checked sometimes and rethrown is not `null`).